library/core/src/any.rs

```diff
@@ -700,3 +778,252 @@ pub const fn type_name<T: ?Sized>() -> &'static str {
 pub const fn type_name_of_val<T: ?Sized>(_val: &T) -> &'static str {
     type_name::<T>()
 }
+
+///////////////////////////////////////////////////////////////////////////////
+// Provider trait
+///////////////////////////////////////////////////////////////////////////////
+
+/// Trait implemented by a type which can dynamically provide tagged values.
+#[unstable(feature = "provide_any", issue = "none")]
+pub trait Provider {
+    /// Object providers should implement this method to provide *all* values they are able to
+    /// provide using `req`.
+    #[unstable(feature = "provide_any", issue = "none")]
+    fn provide<'a>(&'a self, req: Requisition<'a, '_>);
+}
+
+/// Request a specific value by a given tag from the `Provider`.
+#[unstable(feature = "provide_any", issue = "none")]
+pub fn request_by_type_tag<'a, I>(provider: &'a dyn Provider) -> Option<I::Type>
+where
+    I: TypeTag<'a>,
+{
+    let mut req: ConcreteRequisition<'a, I> = RequisitionImpl { tagged: TagValue(None) };
+    provider.provide(Requisition(&mut req));
+    req.tagged.0
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Type tags and the Tagged trait
+///////////////////////////////////////////////////////////////////////////////
+
+/// This trait is implemented by specific `TypeTag` types in order to allow
+/// describing a type which can be requested for a given lifetime `'a`.
+///
+/// A few example implementations for type-driven `TypeTag`s can be found in the
+/// [`tags`] module, although crates may also implement their own tags for more
+/// complex types with internal lifetimes.
+#[unstable(feature = "provide_any", issue = "none")]
+pub trait TypeTag<'a>: Sized + 'static {
+    /// The type of values which may be tagged by this `TypeTag` for the given
+    /// lifetime.
+    #[unstable(feature = "provide_any", issue = "none")]
+    type Type: 'a;
+}
+
+#[unstable(feature = "provide_any", issue = "none")]
+pub mod tags {
+    //! Type tags are used to identify a type using a separate value. This module includes type tags
+    //! for some very common types.
+    //!
+    //! Many users of the provider APIs will not need to use type tags at all. But if you want to
+    //! use them with more complex types (typically those including lifetime parameters), you will
+    //! need to write your own tags.
+
+    use super::TypeTag;
+    use crate::marker::PhantomData;
+
+    /// Type-based `TypeTag` for `&'a T` types.
+    #[derive(Debug)]
+    #[unstable(feature = "provide_any", issue = "none")]
+    pub struct Ref<T: ?Sized + 'static>(PhantomData<T>);
```

Comment:
I love how the ui tests broke because you introduced a `Ref` type that made it so there are two available `Ref` types which disabled the compiler's output shortening so that now it outputs the full path of `Ref` in the ui test diagnostics.